video/out/vo_tct.c

```diff
@@ -0,0 +1,251 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <libswscale/swscale.h>
+
+#include "options/m_option.h"
+#include "config.h"
+#include "vo.h"
+#include "sub/osd.h"
+#include "video/sws_utils.h"
+#include "video/mp_image.h"
+
+#define IMGFMT IMGFMT_BGR24
+
+#define ALGO_PLAIN 1
+#define ALGO_HALF_BLOCKS 2
+
+#define ESC_HIDE_CURSOR "\e[?25l"
+#define ESC_RESTORE_CURSOR "\e[?25h"
+#define ESC_CLEAR_SCREEN "\e[2J"
+#define ESC_CLEAR_COLORS "\e[0m"
+#define ESC_GOTOXY "\e[%d;%df"
+#define ESC_COLOR_BACKGROUND "\e[48;2;%d;%d;%dm"
+#define ESC_COLOR_FOREGROUND "\e[38;2;%d;%d;%dm"
+
+struct priv {
+    int algo;
+    size_t buffer_size;
+    char *buffer;
+    int swidth;
+    int sheight;
+    struct mp_image *frame;
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_sws_context *sws;
+};
+
+static size_t write_plain(
+    const int dwidth, const int dheight,
+    const int swidth, const int sheight,
+    const unsigned char *source, const int source_stride,
+    char *target, const size_t target_size)
+{
+    char *target_start = target;
+    const int tx = (dwidth - swidth) / 2;
+    const int ty = (dheight - sheight) / 2;
+    for (int y = 0; y < sheight; y++) {
+        const unsigned char *row = source ? source + y * source_stride : NULL;
+        target += snprintf(target, target_size, ESC_GOTOXY, ty + y, tx);
+        for (int x = 0; x < swidth; x++) {
+            unsigned char b = row == NULL ? 255 : *row++;
+            unsigned char g = row == NULL ? 255 : *row++;
+            unsigned char r = row == NULL ? 255 : *row++;
+            target += snprintf(target, target_size, ESC_COLOR_BACKGROUND, r, g, b);
+            target += snprintf(target, target_size, " ");
+        }
+        target += snprintf(target, target_size, ESC_CLEAR_COLORS);
+        target += snprintf(target, target_size, "\n");
+    }
+    return target - target_start;
```

Comment:
I'm not really using `target_size` to avoid overflows - my primary goal was to have the same code calculate the buffer size (when it's passed `target=NULL` and `target_size=0`) and actually print to the buffer with assumption "it's big enough". This is because I wanted to keep it in one function, because having two functions - one for measuring buffer size and another for actual printing - would need to duplicate all the conditions, for loops etc inside and this introduces maintenance overhead.